src/data/W/cardinal.lean

```diff
@@ -0,0 +1,89 @@
+/-
+Copyright (c) 2021 Chris Hughes. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Hughes
+-/
+import set_theory.cardinal_ordinal
+import data.W.basic
+/-!
+# Cardinality of W-types
+
+This file proves some theorems about the cardinality of W-types. The main result is
+`cardinal_mk_le_max_omega_of_fintype` which says that if for any `a : α`,
+`β a` is finite, then the cardinality of `W_type β` is at most the maximum of the
+cardinality of `α` or `cardinal.omega`.
+This has applications in first order logic, and can be used to prove that in the
+cardinality of the set of terms in a language with is bound by the cardinality
+of the set of function and constant symbols if that set is infinite, and the set of terms is
+countable if there are finitely many constant and function symbols.
+
+## Tags
+
+W, W type, cardinal, first order
+-/
+universe u
+
+variables {α : Type u} {β : α → Type u}
+
+noncomputable theory
+
+open cardinal
+
+namespace W_type
+
+lemma cardinal_mk_eq_sum : cardinal.mk (W_type β) =
+  cardinal.sum (λ a : α, cardinal.mk (W_type β) ^ cardinal.mk (β a)) :=
+begin
+  simp only [cardinal.lift_mk, cardinal.power_def, cardinal.sum_mk],
+  exact cardinal.eq.2 ⟨equiv_sigma β⟩
+end
+
+lemma cardinal_mk_le_of_le {κ : cardinal.{u}}
+  (hκ : cardinal.sum (λ a : α, κ ^ cardinal.mk (β a)) ≤ κ) :
+  cardinal.mk (W_type β) ≤ κ :=
+begin
+  conv_rhs { rw ← cardinal.mk_out κ},
+  rw [← cardinal.mk_out κ] at hκ,
+  simp only [cardinal.power_def, cardinal.sum_mk, cardinal.le_def] at hκ,
+  cases hκ,
+  exact cardinal.mk_le_of_injective (to_type_injective _ hκ.1 hκ.2)
+end
+
+/-- If, for any `a : α`, `β a` is finite, then the cardinality of `W_type β`
+  is at most the cardinality of `α` or  -/
+lemma cardinal_mk_le_max_omega_of_fintype [Π a, fintype (β a)] : cardinal.mk (W_type β) ≤
+  max (cardinal.mk α) omega :=
```

Comment:
Could we have a version of this theorem using `countable` or `denumerable`?

Reply:
There already is one using `encodable` by Jeremy.